Trying to explain the feature concepts first (#70)
WICG/shared-element-transitions: explainer.md

# Design

The goal is to provide a mechanism and API which will allow simple transitions like above to be specified in CSS, building on CSS animations, but also allow for more complex transition to be performed via JavaScript, building on the Web Animations API.

This section covers the concepts and mechanisms, while a later section looks at possible API shapes.

Performing a transition from Page-A to Page-B requires parts of both to be on screen at the same time, potentially moving independently. This is currently impossible in a cross-document navigation, but it's still hard in an SPA (single page app) navigation. You need to make sure that the outgoing state can't receive additional interactions, and ensure the presence of both states doesn't create a confusing experience for those using accessibility technology.

The aim of this design is to allow for representations of both Page-A and Page-B to exist at the same time, without the usability, accessibility, or memory concerns of having both complete DOM trees alive.

Here's the example that will be used to explain the design:

<img alt="Page-A and Page-B" src="media/pages.png?raw=true">

> The transition between Page-A and Page-B can be a full cross-document navigation between two same-origin pages, or it can be an SPA navigation. The main mechanism and concepts are the same between the two.

Reviewer: This seems repetitive with line 26?

Author: I'll try to reword it. I'm trying to say "these concepts apply to both MPA and SPA, although the API may differ"

# Use-Cases

A visual demo of some example transition patterns targeted by this feature are [here](https://material.io/design/motion/the-motion-system.html#transition-patterns). The following is a summary of the semantics of these transition patterns:

* Root Transitions: The full page content animates between two web pages with an optional static UI element on top. Examples 1 & 2 [here](https://material.io/design/motion/the-motion-system.html#shared-axis) are demonstrations of this.
* Shared Element to Root Transitions: A persistent UI element morphs into the full page content on the next web page. [Container transform](https://material.io/design/motion/the-motion-system.html#container-transform) shows an example.
* Shared Element Transitions: A persistent UI element morphs into another UI element on the next web page. The element's contents and shape can change during this transition. This [video](https://www.youtube.com/watch?v=SGnZN3NE0jA) shows an example.
* Entry/Exit Transitions: A UI element animates as it exits or enters the screen. This [issue](https://github.com/WICG/shared-element-transitions/issues/37) shows an example.

# Design

> The goal is to provide a mechanism and API which will allow simple transitions like above to be specified in CSS, building on CSS animations, but also allow for more complex transition to be performed via JavaScript, building on the Web Animations API.

Author: What do you mean by "animations sticking to the finger"?

> The goal is to provide a mechanism and API which will allow simple transitions like above to be specified in CSS, building on CSS animations, but also allow for more complex transition to be performed via JavaScript, building on the Web Animations API.

Reviewer 2: By sticking to the finger, I mean animations that track touch input. So I can slide something 30% of the way and hold my finger there, then slide it another 10% and the animation would advance another 10%. As opposed to what we're doing here in that once the animation started, it will run and finish without the user being able to affect it.

> The goal is to provide a mechanism and API which will allow simple transitions like above to be specified in CSS, building on CSS animations, but also allow for more complex transition to be performed via JavaScript, building on the Web Animations API.

Reviewer 3: "Is it a non-goal to allow simple transitions easily (ie with no custom CSS)?" This seems to imply that defaults are a non-goal. I don't think that's right.

The aim of this design is to allow for representations of both Page-A and Page-B to exist at the same time, without the usability, accessibility, or memory concerns of having both complete DOM trees alive.

Here's the example that will be used to explain the design:

<img alt="Page-A and Page-B" src="media/pages.png?raw=true">

The transition between Page-A and Page-B can be a full cross-document navigation between two same-origin pages, or it can be an SPA navigation. The main mechanism and concepts are the same between the two.

Cross-origin transitions are something we want to tackle, but may have significant differences and restrictions for security reasons. Cross-origin transitions are not covered in this document.

## Part 1: The offering

> Before Page-A goes away, it offers up parts of itself to be used in the transition. Generally, this will mean one part per 'thing' that acts independently during the transition. For the example transition, the parts are:

Reviewer 4: Re-iterating from the internal discussion, the way we identify "parts" of a page that are cached and retained is through elements. So I find the "part" terminology unnecessarily abstract. We should just call them elements and it would make sense to anyone familiar with web elements.

We started with the word shared because I guess this was originally being used for transitions where we have matched elements from both ages. So they were sort of "shared" between the 2 pages. Also because we borrowed from the Android terminology of the same feature. I'd prefer to stick with shared for now since it's used consistently in all other documentation/code and we don't have any better ideas. If this comes up during external discussion, we can field suggestions and change it for once.

- The header
- The share button
- The rest (referred to as the page root)

https://user-images.githubusercontent.com/93594/141104275-6d1fb67a-2f73-41e4-9cef-14676798223b.mp4

> Aside from the root, an element can only be offered as a transition part if it has paint containment, which clips child content to the content box, although this can be expanded with `overflow-clip-margin`. The is a compromise made to greatly simplify implementation.

Reviewer 4: Sure. You can move the "ease of implementation" part to the first one which talks about contain:paint.

The current element() spec is trying to work around the fragmentation issue so I thought it was implementation complexity : "Elements broken across lines or columns are just rendered with their decorated bounding box." But I don't completely understand what this means.

Additionally, the element must be a single rect. As in, it doesn't break across lines or columns.

> - Open question: If the developer tries to break the above rules, what should happen? Is paint containment applied by the browser? Is that element dropped from the transition? Is the whole transition abandoned?

Reviewer 2: Yeah, I think that's what I'm saying. I mean out of the two choices:

* If we require paint containment, it means that we require the developer to do some work before sharing the element. Then when the transition happens, we either 1. issue a warning if containment was not satisfied and skip it, or 2. ensure that element looks the same way during the transition as before
* If we automatically do paint containment, it means it will work automatically but 1. we still have to issue a warning if we can't do a containment (such as non-atomic inline - `<span>` for example) and 2. things may temporarily shift out of place and be clipped differently for the duration of the transition

I prefer the first option, but I guess that's why it's an open question :)

When a developer offers part of the page for transitions, there are two modes they can choose from:

> ### As a single texture

Reviewer 5: It's a lot like a [css image](https://drafts.csswg.org/css-images-4/#image-values) though.

> ### As a single texture

Author: I'll go with "CSS image"

> The entire painting of the element is captured, including things which appear outside of its bounding box such as shadows and blurs, as a single texture at device-pixel resolution.

Reviewer 4: Yup. We can use a vector graphic instead of bitmaps/textures.

> The entire painting of the element is captured, including things which appear outside of its bounding box such as shadows and blurs, as a single texture at device-pixel resolution.

Reviewer 4: The reasons I suggested image :

- It's essentially a vector graphic if we're caching a display list. Vector graphics like SVGs are also referenced as image.
- The output produced by element() is also called an image.

https://user-images.githubusercontent.com/93594/141118353-d62d19a1-0964-4fa0-880f-bdde656ce899.mp4

> - Open question: Which effects from a parent element are baked into this texture? E.g. things like `filter`, `opacity`, `transform`.

Reviewer 4: We capture it at 0deg and reapply the transform to the container. The transform we're using currently is a squashed matrix which includes a translation for the element's layout position + transforms from the ancestor chain. Ian actually brought up a good point that we should use a list of transform operations instead of a single matrix so 3-d transforms work correctly (https://chromium-review.googlesource.com/c/chromium/src/+/3271751/1..4/third_party/blink/renderer/core/document_transition/document_transition_container_element.cc#b154).

But in either case, it will be a transform on the container and won't apply to the image.

Capturing as a texture avoids the interactivity risks, complexities, and memory impact of fully preserving these parts of Page-A as live DOM.

The root is always captured in this way, and is also clipped to the viewport, as capturing the entire page would take an enormous amount of memory in some cases.

- Open question: Should we have a way to expand this area for particular transitions? For example, transitions that involve vertical movement?
- Open question: Other elements can also be massive. Do we need a way to limit and control the captured size of those?

> This mode works great for the share button and the root, as their transitions can be represented by simple transforms. However, the header changes size without stretching its shadow, and the content of the header moves independently. There's another mode for that:

Reviewer 4: I don't think talking about element() is getting into API, it's about implementation. When we talk about representing the element as an image, it helps to clarify that we don't need to define that behaviour from scratch. The modifications we're suggesting are minor and make sense even outside the context of this feature.

> This mode works great for the share button and the root, as their transitions can be represented by simple transforms. However, the header changes size without stretching its shadow, and the content of the header moves independently. There's another mode for that:

Reviewer 4: Let's decide on calling it something with an explanation when the term is introduced. Since we also want to use that term for the replaced element that's displaying this content. I'm comfortable with image since what we cache can be a vector graphic or rasterized pixels and there is precedent for using "image" for both. This is aside from the element() aspect.

And, okay if it seems like this is too early in the doc to say this is element(). As long as when we do mention it, we're indicating that this is the confident option we're advocating for and not an open question.

> This mode works great for the share button and the root, as their transitions can be represented by simple transforms. However, the header changes size without stretching its shadow, and the content of the header moves independently. There's another mode for that:

Reviewer: It is not too early to say element(). If we are not specific about a proposal the CSSWG will not know what to review.

That's why when we go to them we need to have a precise thing that is being proposed, which should be:

* element() for getting images of things, with list of edits to element()
* top layer for what it means to be "on top"
* pseudo elements for author styling and UA stylesheet styling

Those are the key things we need to go to them with for alignment.

> This mode works great for the share button and the root, as their transitions can be represented by simple transforms. However, the header changes size without stretching its shadow, and the content of the header moves independently. There's another mode for that:

Reviewer: The key thing for the discussion with them is to map everything into CSS spec concepts, to make clear that this is not a scary new thing but builds incrementally on existing things in a layered way.

> This mode works great for the share button and the root, as their transitions can be represented by simple transforms. However, the header changes size without stretching its shadow, and the content of the header moves independently. There's another mode for that:

Reviewer: This does not need to be the exposed API, it's the CSS concepts that explain the entire thing. The purpose of spelling out style sheets and element() in the UA style sheet is proof that the layering works.

> This mode works great for the share button and the root, as their transitions can be represented by simple transforms. However, the header changes size without stretching its shadow, and the content of the header moves independently. There's another mode for that:

Reviewer 2: Sounds good.

> element() for getting images of things, with list of edits to element()

Other than the to paint the border or just the contents type of thing, I wanted to mention one other thing to call out about element(). The element() function allows selecting unrendered paint sources (like canvas with display:none). We shouldn't share/render these.

For rendered elements, it lists `an element that is rendered, is not a descendant of a replaced element, and generates a stacking context` as being able to be captured. I'm not sure why the descendant of a replaced element clause is there. It'd be nice to find out. I know things like canvas can have fallback content, but I'm not entirely sure in which situations do we render it

> This mode works great for the share button and the root, as their transitions can be represented by simple transforms. However, the header changes size without stretching its shadow, and the content of the header moves independently. There's another mode for that:

Author: I'm going to add a quick reference to `element()` earlier to reassure the reader that this isn't a totally new concept, but I'll also signal that it's not completely the same and the differences are discussed later.

> ### As a container + child texture

Reviewer 4: SGTM.

> In this mode the computed styles of the element are copied over, so they can be re-rendered beyond just transforming a texture. This allows the developer to animate properties such as width and height, borders, background color… anything that can be animated with CSS. It also allows the developer to create animations that clip child content, via `overflow: hidden`.

Reviewer 4: "It isn't a subset right?" : You can interpolate the value of a property on the container but it won't have any effect if the property affects the element's content that has been baked into the image. That's why I brought up font-size as an example. With the regular element, you'd expect the text to resize and re-layout as the value interpolates. But in this mode, changing it would be a no-op.

So for this mode, we'd end up having to define which properties are not painted into the image and can be animated.

The children of the element (including pseudos and text nodes) are captured into a single texture that can be animated independently.

https://user-images.githubusercontent.com/93594/141118395-8d65da49-a5ab-41c6-8458-917e55d4b77b.mp4

> A mode like this is unnecessary complexity for the share button in the example transition, but gives the developer the freedom they need for the header transition.

Reviewer 4: Yup, my bad. Fixed.

> - Open question: Do we need this for 'v1'?

Reviewer: s/this/the second mode where styles are copied to a container element/

### Nested parts

> In the example transition, the content of the header cross-fades from Page-A to Page-B. An even smoother transition could be achieved by also animating the site title and avatar 'chip' independently. However, to reduce the scope of 'v1', offering parts nested in other offered parts is not allowed.

Reviewer 2: I feel like this kind of makes the 'root' element special in that it's kind of like a shared element that can have other shared elements nested in it. I guess that's by design.

The question I had is whether we should call out (later in the API) what do we propose happens if the developer does specify nested elements as shared. Is this something that we're going to verify at offer up phase or at transition phase or both?

> In the example transition, the content of the header cross-fades from Page-A to Page-B. An even smoother transition could be achieved by also animating the site title and avatar 'chip' independently. However, to reduce the scope of 'v1', offering parts nested in other offered parts is not allowed.

Author: Ohh good question

> - Open question: If the developer tries to do this, what should happen? Is the nested offering ignored? Is the whole transition abandoned?

Reviewer 2: Also when should we verify this? At offer up phase and at transition phase on the new page? What happens if the relationship changes while the transition is happening

## Part 2: The preparation

> The state changes over to Page-B, and Page-A is gone aside from the parts it offered.

Reviewer 4: I said navigation to highlight that during the navigation sequence the browser starts to fetch page-B before page-A is hidden. And at some point the browser decides to hide page-A and show page-B instead. This state will need to be cached between fetching page-B and hiding page-A.

I had originally talked about events from an MPA perspective (navigation, load, ready for first render) and then mentioned that we'll need APIs for developers to provide the equivalent events explicitly for SPA. But can call this "when a transition is initiated" if you wanted to avoid MPA specific terms here.

### Setting the stage

> A fixed-position, viewport-filling overlay is automatically created in the top level.

Reviewer: s/top level/top layer/ throughout.

> A fixed-position, viewport-filling overlay is automatically created in the top level.

Reviewer 4: The initial explainer modeled it as a list of fixed positioned container elements (for each shared element) that draw in a viewport filling overlay (the top-layer). I think you're suggesting another fixed position element above them. Why?

> A fixed-position, viewport-filling overlay is automatically created in the top level.

Reviewer 2: Does this mean we're explicitly not using top layer as defined in the fullscreen spec? It's not clear to me where this layer is hosted. Is it in the DOM?

> A fixed-position, viewport-filling overlay is automatically created in the top level.

Author: I think the intent is to use the top layer, but I'm not as familiar with the details there. We'd still need to define the interaction with fullscreen and dialog.

> A fixed-position, viewport-filling overlay is automatically created in the top level.

Author: > I think you're suggesting another fixed position element above them. Why?

eh, I guess it doesn't matter. I'll stick with your way.

> The offered parts from Page-A are absolutely positioned at (0,0), and moved into their previous viewport-relative positions using a transform.

Reviewer: viewport positioned, not asbpos.

> The offered parts from Page-A are absolutely positioned at (0,0), and moved into their previous viewport-relative positions using a transform.

Reviewer 4: The effect is equivalent indeed but I don't understand the rationale for a fixed position container on top of the elements?

> This is done without the user seeing Page-B. Ideally, there won't be a visual change from the final rendering of Page-A. However, the non-root parts will always render on top of the root, whereas the original page may have drawn other things on top of the offered parts.

Reviewer 4: I didn't think I changed the meaning of this sentence (in terms of like something with exceptions). I was trying to use clearer terms : "non-root parts" -> "shared elements", "original page may have drawn other things on top of the offered parts" -> "shared element was occluded by another element".

> This is done without the user seeing Page-B. Ideally, there won't be a visual change from the final rendering of Page-A. However, the non-root parts will always render on top of the root, whereas the original page may have drawn other things on top of the offered parts.

Reviewer 2: Is the relative paint order preserved for shared elements? AFAIK z-index is ignored in the (fullscreen spec's) top layer and we instead rely on paint order. Or do we account for the z-index before promoting the elements to top layer?

> This is done without the user seeing Page-B. Ideally, there won't be a visual change from the final rendering of Page-A. However, the non-root parts will always render on top of the root, whereas the original page may have drawn other things on top of the offered parts.

Reviewer 5: I would assume that we would order them in the global paint order from their original context, preserving their relative z-index orders where applicable. Note that this does not require preserving the z-index value, just that they would paint in the same order per [spec](https://drafts.csswg.org/css2/#z-index:~:text=Within%20each%20stacking%20context,%20the%20following%20layers%20are%20painted%20in%20back-to-front%20order) which could be as simple as the elements are in that dom order. Note that the order within the target page for associated elements wouldn't be correct in this case presuming that on the target page the elements could have a different order. Maybe z-index should be animated between the two resulting in a 50% flip?

### How are these parts represented?

> Transition parts are represented as elements with the following nesting:

Reviewer 4: Does that read better? I guess I'm trying to say that the state we've preserved (which is image + computed properties/styles for shared elements) is represented using elements.

```

> transition part container

Reviewer 4: Oh, you can leave it as transition container. This was only meant to avoid using "part".

> transition part container

Author: Ah cool. I want to avoid things that might suggest this is an element name or a class name.

└─ transition part
   └─ texture
```

> - **transition part container**: If the part is created as a "container + child texture", this element will have a width and height of the content area of the original element, and have its computed styles reapplied. If the part is created as a "single texture", this element will have a width and height of the border box of the original element.

Reviewer 4: We can mention that the following CSS comes from the UA before the code block? It should be okay since we haven't gotten to any developer code in the explainer up till this point.

The current way felt like reading code in prose but I'm okay either way.

> - **transition part container**: If the part is created as a "container + child texture", this element will have a width and height of the content area of the original element, and have its computed styles reapplied. If the part is created as a "single texture", this element will have a width and height of the border box of the original element.

Reviewer 2: I would at least say "content box" instead of "content area" since the latter makes it sound like an area expanded by overflow-clip-margin, but I don't think that's what we mean right?

- **transition part**: This element has a width and height of 100%, and `isolation: isolate`. This wrapper is useful when cross-fading textures (documented later).
- **texture**: This contains the texture, which may paint outside the parent elements. This would behave like a replaced element, so would work with CSS properties like `object-fit`.

>   - Open question: How is painting outside handled? This element could overflow the parents in a developer-visible way, or we could avoid exposing that part somehow (as we currently do with paint overflowing).

Reviewer 4: We talked about alternatives that we considered but picked one which we're confident about. And then the ones where it's still unclear. To me this falls in the former category, we don't have any argument to avoid exposing it.

That's why I preferred mentioning doing it in the standard developer visible way with a comment mentioning the other choice and why it was dropped.

> - Open question: What are these elements? Pseudo-elements are tricky to handle in some APIs (e.g., no `getBoundingClientRect` equivalent), and the nesting of pseudo-elements seems new. Alternatively, they could be `div`s in a shadow root, which means developers can use regular APIs on these elements.

Reviewer: yes

> - Open question: How are styles applied to these elements? Inline styles are simple, but tricky to override in a stylesheet. An alternative would be to generate a `<style>` and put it, and the transition elements, in a shadow root.

Reviewer 4: Ok. That's a good open question then. Especially when we talk about developers overriding this. It's not just that the transform is supplied by the UA, it's also that its updated to stay in sync with the shared element on the new page. So what happens if a developer overrides this, does the UA clobber the override? And we don't have a concrete answer on how that should be set up.

### Mixing in parts of Page-B and associating them with Page-A parts

> At this stage, Page-B identifies parts of its own page to be involved in the transition. This happens in the same way as part 1 with one difference: The textures and styles from Page-B will be updated if the underlying page updates. This means things like animated gifs will play, rather than being frozen on whatever frame they were on when they were captured.

Reviewer 2: I haven't thought about gifs being frozen, it may be a thing to call out when we're talking about capturing content during the offer up phase. Not sure if that's clear

> The developer can associate particular parts from Page-A to parts from Page-B. This would usually be done if they're equivalent. In this case, the headers, share buttons, and roots are equivalent. When this happens, the texture from the Page-B capture is added to the transition part:

Reviewer 2: > This would usually be done if they're equivalent

Would we still support shared element becoming the root element?

Author: I'm not sure what you mean?

Reviewer 2: The text here suggests that headers transition to headers, buttons to buttons, roots to roots. My question is do we support a div becoming a new root? Ie it's just a plain old element on page-A and it's the root element on page-B [eg](https://material.io/design/motion/the-motion-system.html#container-transform)

```
transition part container
└─ transition part
   ├─ texture (Page-A)
   └─ texture (Page-B)
```

> This allows for the container to be moved as one, while cross-fading the Page-A and Page-B content. The developer will also have access to both computed styles of the container for both Page-A and Page-B, so those can also be transitioned.

Reviewer 4: Hmmm, fair point. The idea I was trying to get across is that the developer should have access to the shared element state being replicated and that depends on the mode. Updated to make that clearer.

Parts don't need to be associated with another part, which allows for transitions involving elements that are only in Page-A or only in Page-B.

The root parts of each page are automatically associated.

## Part 3: The transition

Everything is now in place to perform the transition. The developer can move the parts around using the usual APIs, such as CSS and web animations.

https://user-images.githubusercontent.com/93594/141100217-ba1fa157-cd79-4a9d-b3b4-67484d3c7dbf.mp4

## Part 4: The end

> When the transition is complete, the top level is removed, revealing the real Page-B.

Reviewer 4: I was trying to say that we remove all the elements (container, wrapper, image) that have been created for the transition. But perhaps since they are in the "overlay", saying that it's removed is the same thing. I'm still unsure about what this overlay is supposed to be.

If we decide that the "overlay" is top layer then this can be reworded to removing UA created elements from the top layer.

> When the transition is complete, the top level is removed, revealing the real Page-B.

Reviewer 2: I'd also try and rephrase "revealing the real Page-B". You could always see the real page B, just with the overlay described here, but maybe it's fine :)

# The API

> There are a lot of open questions around the API, but this section will give a flavour of the direction.

Reviewer 2: ```suggestion
There are a lot of open questions around the API, but this section will give a flavor of the direction.
```

Although I'm Canadian, I think we stick to American spelling 😂

Little thought has been giving to naming in these APIs, which will be improved when it's clearer what the right direction is.

> Additionally, this API focused on cross-document transitions.

Reviewer 2: ```suggestion
Additionally, this API focused on cross-document transitions, since it is not currently possible with existing web APIs.
```

Author: I meant that the SPA stuff is in a later section. I'll make this clearer.

## Creating a transition part

There are a few ideas for how a developer could offer an element as a transition part.

### Attributes

```html
<header pagetransitiontag="header" pagetransitioncapture="container-and-child">…</header>
```

> Here, the presence of the `pagetransitiontag` attribute marks this element as a transition part, and the `pagetransitioncapture` attribute is used to indicate how it's captured (as a single texture vs as a container + child texture).

Reviewer 5: It seems like the page-transition-tag css property is strictly superior to the pageTransitionTag attribute, do we need both? If we have both what does it mean if they have different values?

Author: Yeah, it's never going to be both. We need to pick one. I prefer CSS properties too.

Multiple elements could have the same `pagetransitiontag` value, which is useful for aggregate content such as comments.

Items on Page-A and Page-B which have the same `pagetransitiontag` are automatically associated with each other. If multiple items have the same value, they're associated in DOM order.

### CSS properties

```css
.header {
  page-transition-tag: header;
  page-transition-capture: container-and-child;
}
```

This operates similar to the attributes, but it's easier to give multiple elements the same tag. It's also easier to change values depending on viewport (`@media`) and browser support (`@supports`).

### JS API

```js
// In Page-A
addEventListener('navigate', (event) => {
  event.prepareSameOriginPageTransition((transition) => {

>     transition.offerItem('header', document.querySelector('.header'), {

Reviewer 4: If you want a |data| per offered element then we'd need something like setData('header', data) but it doesn't need to also be a way to specify offered elements.

I also feel that a single setData API is enough, having it per offered element seems redundant. It's already an |any| type, the developer can pass a dictionary if they want to key things per tag.

>     transition.offerItem('header', document.querySelector('.header'), {

Author: I'll drop the per-element data thing. Just using DOM is interesting. I'll add an open question.

      data: {…},
      capture: 'container-and-child',
    });
  });
});
```

Expanding on the capabilities of the CSS properties, a JS API allows the developer to change which parts are offered depending on the destination of the navigation, and the direction of navigation (back vs forward). This builds on top of the [app-history API](https://github.com/WICG/app-history).

Also, `data` can be associated with the element. This can be anything structured-clonable, and will be made available to Page-B.

```js
// In Page-B
document.performTransition((transition) => {
  const pageAHeader = transition.offeredItems.get('header');
  console.log(pageAHeader.data);

>   const pageBHeader = transition.createItem(document.querySelector('.header'), {

Reviewer 4: Yeah, you'd have to remove them at the end of the transition. And fair point that the JS API can be limited to the particular transition.

But I think this ties into my other comment about not coupling with app-history. If you wanted that to be "tied" to a particular navigation then the offered elements have to be associated to a navigation event somehow. And I didn't want to add a navigation event to the documentTransition API since app-history is doing that already. But also not tightly couple with it. That's why I was hoping the semantics to be "this element is offered(old page)/used(new page) for whatever the next transition is" from the documentTransition API perspective. And developers relying on other events to set it based on the particular situation.

Maybe we can leave this as an open question and come back to it when we narrow down API details?

    capture: 'container-and-child',
  });

>   transition.matchItems(pageAHeader, pageBHeader);

Reviewer 4: My motivation for the ID being unique was so :

- we could have a purely declaratively way to match elements.
- target a unique element using this ID in script APIs (like animate(..), getComputedStyle(..)).

If there is an alternate way to do this with multiple items having the same tag, I'd be fine with that too.

If an item exists in Page-B only, it could have a default animation that takes its container from opacity 0 to 1.

If an item exists in both Page-A and Page-B, and both are 'container and child', it could have a default animation that takes its container from Page-A styles to Page-B styles (which will include the transform used for positioning), while cross-fading the two textures.

If an item exists in both Page-A and Page-B, and neither are 'container and child', it could have a default animation that takes its container from Page-A size and position to Page-B styles via a transform, while cross-fading the two textures.

- Open question: What if the Page-A item is 'container and child' but the Page-B item is 'single texture'?

Because the textures are sized to 100% of the container, the textures will also change size throughout the transition. How these are scaled can be changed using regular CSS features like `object-fit`.

In all cases, the duration and easing is some undecided default, that could even be platform independent.

- Open question: When will the automatic animation start? When the browser would usually first render Page-B?

> - Open question: Automatic animations work well for things which are at least partially in-viewport in both Page-A and Page-B, but it gets tricky if you consider a non-sticky header that scrolled out of view by 1000s of pixels.

Reviewer 4: Ok. SGTM then.

### CSS animation

CSS can be used to build on automatic animations, or override the default.

```css
::page-transition-container(header) {
  /* … */
}
```

Element selectors:

- `::page-transition-container(name)` - Select the transition part containers of a given `page-transition-tag`.
- `::page-transition-part(name)` - Select the transition parts of a given `page-transition-tag`.
- `::page-transition-texture-outgoing(name)` - Select the outgoing texture of a given `page-transition-tag`.
- `::page-transition-texture-incoming(name)` - Select the incoming texture of a given `page-transition-tag`.
- `::page-transition-root-outgoing` - Select the outgoing root texture.
- `::page-transition-root-incoming` - Select the incoming root texture.

> - `::page-transition-background` - This can be set for cases where something needs to be rendered underneath the root textures.

Author: Agreed! I was thinking the same, but messed it up in the API

> - `::page-transition-background` - This can be set for cases where something needs to be rendered underneath the root textures.

Reviewer 4: Ok. So maybe we can just mention that the setup is the same with different selectors (page-transition-root-container/wrapper/image-outgoing/image-incoming) for the root elements. We thought about doing it using a reserved "root" keyword as well but that's just bike-shedding. :)

These will be selecting a pseudo-element, or an element in a UA-created shadow DOM.

```css
::page-transition-container(header) {
  animation-delay: 300ms;
}
```

CSS can be used to make changes to the automatic animation, or override `animation-name` to remove the default.

### Web animation API

```js
// In Page-B
document.performTransition((transition) => {
  // Build up transition parts, then…

>   transition.root.querySelector('[part=header]').animate(…);

Author: If we go for the shadow root solution, yes. The API will look different if we stay with pseudos, and worse in my opinion, but there may be reasons shadow DOM is worse in other cases.

});
```

> - Open question: What's the deadline for calling `performTransition`?

Author: Agreed. This question is purely for the JS API.

> Again, this is a half-baked sketch. This example assumes that elements in the shadow root are given a part attribute, but there's likely some better way to address the elements.

Author: I wonder if there's a way to work around this

> If the 'stage' of the transition is exposed as a shadow root, the developer can interact with the elements in a regular way. The developer could even create elements specifically for the transition.

Author: I was thinking you could add an element to perform some kind of fade-to-back, or a wipe effect, or even add a `<canvas>` and do some really weird shit.

What's the downside to letting devs work with this like regular DOM?

> If the 'stage' of the transition is exposed as a shadow root, the developer can interact with the elements in a regular way. The developer could even create elements specifically for the transition.

Reviewer 4: My main tension comes from the fact that we're mirroring state in the actual DOM in this DOM representation that we're creating in the overlay/top layer. It's cached for old page but "live" for the new page. Right now it's a flat list of images but would include the hierarchy to support nested shared elements.

The more things we let developers modify in this DOM representation, the more difficult it becomes to reason about how that interacts with the browser keeping it in sync and defining how the 2 states should be merged/transitioned.

If we see legit use-cases for fancy stuff from power users, then I'd rather we just provide the primitives being used to build this representation (cached image, computed properties, style from old page and a live image for an element on the new page) directly to the developer and they can create their own overlay/DOM representation. Vlad actually also mentioned something along these lines.

For now let's try to limit the complexity to the use-cases we know. And I think hooks for animating these elements and reading computed info (ComputedStyle, getBoundingClientRect) in script is sufficient.

> If the 'stage' of the transition is exposed as a shadow root, the developer can interact with the elements in a regular way. The developer could even create elements specifically for the transition.

Reviewer: The downside is that it's hard, and potentially dangerous (in the sense of accidentally very hard complexity in implementations, but also possibly security issues), to let developers do anything. For example, if we let them stick arbitrary stuff in the top layer whenever they want, then the top layer loses its value as a special place jealously guarded by the UA. Then we'd have to say how all this stuff is reflected in the DOM, mutation observers, holding onto references for elements which lost their cached backing, etc. If we expose element() to arbitrary content then they can create circularities in the rendering pipeline/tree ordering.

This is also why I'm in favor of pseudo elements instead of shadow DOM. Pseudo elements don't exist in the DOM, they only exist in the render box tree. Likewise, the top layer is a render box tree concept, not a DOM one. This has a lot of advantages, because it makes it possible for the UA to lift stuff into fullscreen without disturbing or confusing virtual DOM frameworks, event listeners, or other script stuff that operates on the DOM. Likewise, with this API, we can use the power of the render box tree and the UA implementation to elegantly lift stuff out of the DOM into the top layer, and also allow the developer to customize it (just like they can do for fullscreen and dialogs) via style sheets. And based on UA timeouts, or user overrides, we can end the animations when we want, or even skip them entirely, without the developer having to do anything (or being able to do anything, in the case of the user overriding).

If we find tons of demand for this feature for advanced modes where there are different topologies required, that's a great problem to have and we can certainly build augmented versions of this API in the future to meet those needs. In the meantime, we should go with a solution that elegantly solves the know problem and fits into existing concepts and implemented features of browsers as much as possible.

## Signaling the end of a page transition

At the start of the transition, the browser could gather all the `Animation`s active on the stage, and assume the animation is complete once all animations finish.

In addition, the JS API could include a way to provide a promise which keeps the transition active, allowing for animations driven some other way, such as `requestAnimationFrame`.

## Example

Using the sketches above, here's how the example Page-A to Page-B transition could be done:

> https://user-images.githubusercontent.com/93594/141100217-ba1fa157-cd79-4a9d-b3b4-67484d3c7dbf.mp4

Reviewer 4: Can you also add an example for what this would look like if the capture was in single image mode?

Author: Ohh, shall do. I think it'd make more sense earlier in the document, just before the section on the other mode.

In Page-A:

```css
.header {
  page-transition-tag: header;
  page-transition-capture: container-and-child;
}

.share-button {
  page-transition-tag: share-button;
}
```

In Page-B:

```css
.header {
  page-transition-tag: header;
  page-transition-capture: container-and-child;
}

.share-button {
  page-transition-tag: share-button;
}

/* Slide the roots from right to left */
@keyframes slide-left {
  to { transform: translateX(-100%); }
}

::page-transition-root-outgoing {
  animation-name: slide-left;
}

::page-transition-root-incoming {
  left: 100%;
  animation-name: slide-left;
}

/* Prevent the header content from stretching */
::page-transition-texture-outgoing(share-button),
::page-transition-texture-incoming(share-button) {
  object-fit: cover;
}
```

# Single-Page-App API

The mechanism for cross-document transitions and SPA transitions involves the same phases, so an SPA API will expose those parts in the same page.

Half-baked sketch:

```js
document.documentTransition.prepare(async (transition) => {

>   // …Identify State-A elements to capture, then:

Reviewer 4: I'm assuming the async callback is dispatched *after* the prepare phase finishes and the next scene can be asynchronously loaded. So identifying state-A elements can't be in this callback.

Author: In this sketch, it's in state-a before `await updatePageStateSomehow()` is called. I went with this patter for some of the same reasons as https://github.com/WICG/shared-element-transitions/issues/24, but it's just a quick sketch and I haven't really thought it through.

  await updatePageStateSomehow();
  // The page is now in State-B.
  // …Identify State-B elements to capture, then:
  transition.start();
});
```

# Cross-fading

Cross-fading two DOM elements is currently impossible if both layers feature transparency. This is due to the default composition operation: black with 50% opacity layered over black with 50% opacity becomes black with 75% opacity.

> However, the [plus-lighter](https://drafts.fxtf.org/compositing/#porterduffcompositingoperators_plus_lighter) compositing operation does the right thing when isolated to a set of elements whose `opacity` values add to 1. The "transition part" wrapper performs this isolation.

Reviewer 4: ```suggestion
However, the [plus-lighter](https://drafts.fxtf.org/compositing/#porterduffcompositingoperators_plus_lighter) compositing operation does the right thing when isolated to a set of elements whose `opacity` values add to 1. The "transition wrapper" is meant to provide this isolation.
```

Allowing `mix-blend-mode` to be set to `plus-lighter` will enable developers to create real cross-fades between elements for this feature and elsewhere.

> # Relation to `element()`

Reviewer 4: I made a suggestion above to include this within the "single image" mode and specify the modifications needed to element() to support that mode.

CSS has an [`element()`](https://drafts.csswg.org/css-images-4/#element-notation) feature which allows the appearance of an element to be used as an image.

This doesn't quite match either of the cases where we need to capture an element as a texture.

When capturing 'as a single texture', it seems much easier for developers if we expand the capture to include things outside the border box, such as box shadows. `element()` clips at the border box.

> When capturing 'as a container + child texture', we capture the combination of the element's children as texture, clipped to the content box. Again this is different to `element()`.

Reviewer 4: For this also, add it in the "computed style + content" section? I like the idea of introducing the concept and then referring existing concepts which come close to what we want.

However, these variations could be included in `element()` using modifiers or similarly named functions (e.g. `element-children()`).

# Security/Privacy Considerations

The security considerations below cover same-origin transitions.

* Script can never read pixel content in the textures. This is necessary since the document may embed cross-origin content (iframes, CORS resources, etc.) and multiple restricted user information (visited links history, dictionary used for spell check, etc.)

> * If an element is captured 'as a container + child texture', any external resources specified on the container, such as background images, will be re-fetched in the context of the new page to account for differences in sandboxing.

Reviewer 4: ```suggestion
* If an element is captured 'as a ComputedStyle + content texture', any external resources specified in the style of the shared element , such as background images, will be re-fetched in the context of the new page to account for differences in sandboxing. If access to these resources is prevented due to a difference in CORS/COEP policy, then re-fetching such resources will fail.
```

Author: > difference in CORS/COEP policy

It isn't just that right? Would CSP and other sandboxing come into play too?

Reviewer 4: Sure. The main point I was trying to emphasize was that we should call out that re-fetching in the new context can fail. And the developer should ensure not to go for style + content mode for an element which has such resources.

Cross-origin transitions aren't yet defined, but are likely to be heavily restricted.

# Interactivity and accessibility

> Page transitions are a purely visual affordance. In terms of interactivity, transition elements will behave like `div`s regardless of the original element.

Reviewer 4: Should hit-testing ignore the transition elements? I think this is a good open question. It's not clear to me how input routing should work while we're displaying the transition setup. Since the only content that's visible onscreen is the transition elements and developers can't register event handlers on them.

Author: Yeah, the right answer isn't clear to me

This section covers the concepts and mechanisms, while a later section looks at possible API shapes.

Performing a transition from Page-A to Page-B requires parts of both to be on screen at the same time, potentially moving independently. This is currently impossible in a cross-document navigation, but it's still hard in an SPA (single page app) navigation. You need to make sure that the outgoing state persists along with the incoming state, that it can't receive additional interactions, and ensure the presence of both states doesn't create a confusing experience for those using accessibility technology.

The aim of this design is to allow for representations of both Page-A and Page-B to exist at the same time, without the usability, accessibility, performance, security and memory concerns of having both complete DOM trees alive.

Here's the example that will be used to explain the design:

<img alt="Page-A and Page-B" src="media/pages.png?raw=true">

The concepts and process described in this section apply for both MPA and SPA transition, however the API will differ in parts.

> Cross-origin transitions are something we want to tackle, but may have significant differences and restrictions for security reasons. This, cross-origin transitions are not covered in this document.

Reviewer 4: ```suggestion
Cross-origin transitions are something we want to tackle, but may have significant differences and restrictions for security reasons. Cross-origin transitions are not covered in this document.
```

## Part 1: The offering

Before Page-A goes away, it offers up elements to be used in the transition. Generally, this will mean an element that animates independently during the transition. For the example transition, the elements are:

- The header
- The share button
- The rest (referred to as the page root)

https://user-images.githubusercontent.com/93594/141104275-6d1fb67a-2f73-41e4-9cef-14676798223b.mp4

Aside from the root, an element offered for a transition has the following restrictions:

> - [`contain: paint`](https://developer.mozilla.org/en-US/docs/Web/CSS/contain) which ensures that the element is the containing block for all positioned descendants and generates a stacking context. This implies that the child content will be clipped to the context-box but it can be expanded using ['overflow-clip-margin'](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-clip-margin). Being a stacking context and a containing block allows the element to be treated as a single unit, whereas paint containment simplifies implementation.

Reviewer 4: ```suggestion
- [`contain: paint`](https://developer.mozilla.org/en-US/docs/Web/CSS/contain) which ensures that the element is the containing block for all positioned descendants and generates a stacking context. This implies that the child content will be clipped to the context-box but it can be expanded using ['overflow-clip-margin'](https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-clip-margin). Being a stacking context and a containing block allows the element to be treated as a single unit which simplifies implementation.
```

- Open question: Should we have a way to expand the root capture area for particular transitions? For example, transitions that involve vertical movement?
- Open question: Other elements can also be massive. Do we need a way to limit and control the captured size of those?

This mode works great for the share button and the root, as their transitions can be represented by simple transforms. However, the header changes size without stretching its shadow, and the content of the header moves independently and doesn't stretch. There's another mode for that:

### As the element's computed style + content image

In this mode the computed styles of the element are copied over, so they can be re-rendered beyond just transforming an image.

The children of the element (including pseudos and text nodes) are captured into a single CSS image that can be animated independently.

> This allows the developer to animate [animatable CSS properties](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_animated_properties) on the container such as border, background-color, border-radius, opacity. These properties roughly map to the element's box decorations and visual effects. The developer isn't prevents from animating properties like `font-size`, but since the children are captured as an image, it changing the `font-size` won't change the visual output.

Reviewer 4: ```suggestion
This allows the developer to animate [animatable CSS properties](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_animated_properties) on the container such as border, background-color, border-radius, opacity. These properties roughly map to the element's box decorations and visual effects. The developer isn't prevented from animating properties like `font-size` but, since the children are captured as an image, changing the `font-size` won't change the visual output.
```

The restriction avoids the need to preserve the hierarchy of shared elements and associated properties (transform, clip, effects inherited by descendants) in the DOM representation created to render the images referenced above. This helps in minimizing the scope for 'v1' of the feature.

- Open question: If the developer tries to do this, what should happen, and when is that verified? Is the nested offering ignored? Is the whole transition abandoned?

## Part 2: The preparation

At this point the state has changed over to Page-B, and Page-A is gone aside from the elements it offered. In the MPA case, this happens when the navigation is complete. In the SPA case, this happens when the DOM is in the Page-B state and the developer signals that the change is complete (how to make that signal is discussed later in the API).

### Setting the stage

The offered elements from Page-A are fixed position at (0,0) in the [top layer](https://fullscreen.spec.whatwg.org/#top-layer), and moved into their previous viewport-relative positions using the cached transform.

> The overlay ensures that the user continues to see Page-A's visuals as Page-B is loading. Note that this may not reproduce the exact rendering on Page-A. For example, the relative paint order of shared elements is preserved in the overlay. But if a shared element was occluded by another element, the latter is painted into the root's image unless it is also offered as a shared element.

Reviewer 4: ```suggestion
The top layer content ensures that the user continues to see Page-A's visuals as Page-B is loading. Note that this may not reproduce the exact rendering on Page-A. For example, the relative paint order of shared elements is preserved in the top layer. But if a shared element was occluded by another element, the latter is painted into the root's image unless it is also offered as a shared element.
```

> Ideally, there won't be a visual change from the final rendering of Page-A. However, the non-root parts will always render on top of the root, whereas the original page may have drawn other things on top of the offered parts.

Reviewer 4: I thought the sentence above captures this paragraph. Did you still want to keep it?

Page-B is hidden from rendering until the transition is complete.

### How are transition elements represented?

The CSS images and computed properties/styles cached from Page-A are represented as elements with the following nesting:

```
transition element
└─ image wrapper
   └─ image
```

- **transition element**: If the element is created as a "computed style + content image", this element will have a width and height of the content box of the original element, and have its computed styles reapplied. If the part is created as a "single image", this element will have a width and height of the border box of the original element. In either case, this element has a transform applied to position it in viewport space.

> - **image wrapper**: This element has a width and height of 100%, and [`isolation: isolate`](https://developer.mozilla.org/en-US/docs/Web/CSS/isolation). This wrapper is useful when cross-fading textures (documented later).

Reviewer 4: ```suggestion
- **image wrapper**: This element has a width and height of 100%, and [`isolation: isolate`](https://developer.mozilla.org/en-US/docs/Web/CSS/isolation). This wrapper is useful when cross-fading images (documented later).
```

- Open question: How does the outgoing page offer just the root for transition?
- Open question: When is the `performTransition` callback called?

## Defining the animation

How will Page-B define the animation?

### Default animation

The potential default animations setup for the different cases are as follows. Note all of these can be overridden by the developer:

- If an element exists in Page-A only (exit animation), a fade animation takes its container from opacity 1 to 0.
- If an element exists in Page-B only (entry animation), a fade animation takes its container from opacity 0 to 1.
- If an element exists in both Page-A and Page-B, and both are 'container and content', an animation takes the container from Page-A styles to Page-B styles (which will include the transform used for positioning), while cross-fading the two images.
- If an element exists in both Page-A and Page-B, and neither are 'container and child', a transform animation takes its container from Page-A size/transform to Page-B size/transform, while cross-fading the two images.

> - Open question: What if the Page-A element is 'container and child' but the Page-B element is 'single texture'?

Reviewer 4: ```suggestion
- Open question: What if the Page-A element is 'container and child' but the Page-B element is 'single image'?
```

> Because the textures are sized to 100% of the container, the textures will also change size throughout the transition. How these are scaled can be changed using regular CSS features like `object-fit`.

Reviewer 4: ```suggestion
Because the images are sized to 100% of the container, the images will also change size throughout the transition. How these are scaled can be changed using regular CSS features like `object-fit`.
```

In addition, the JS API could include a way to override this by letting the developer provide a promise which keeps the transition active, allowing for animations driven some other way, such as `requestAnimationFrame`.

This is discussed in [#64](https://github.com/WICG/shared-element-transitions/issues/64).

## Example

Using the sketches above, here's how the example Page-A to Page-B transition could be done:

https://user-images.githubusercontent.com/93594/141100217-ba1fa157-cd79-4a9d-b3b4-67484d3c7dbf.mp4

In Page-A:

```css
.header {
  page-transition-tag: header;
  page-transition-capture: container-and-child;
}

.share-button {
  page-transition-tag: share-button;
}
```

In Page-B:

```css
.header {
  page-transition-tag: header;
  page-transition-capture: container-and-child;
}

.share-button {
  page-transition-tag: share-button;
}

/* Slide the roots from right to left */
@keyframes slide-left {
  to { transform: translateX(-100%); }
}

::page-transition-root-outgoing {
  animation-name: slide-left;
}

::page-transition-root-incoming {
  left: 100%;
  animation-name: slide-left;
}

/* Prevent the header content from stretching */

> ::page-transition-texture-outgoing(share-button),

Reviewer 4: ```suggestion
::page-transition-image-outgoing(share-button),
```

> ::page-transition-texture-incoming(share-button) {

Reviewer 4: ```suggestion
::page-transition-image-incoming(share-button) {
```
